File: src/GameServer/test/game/LegacyWatchEquipmentCommandsTest.java
package game;

import assignments.PacketWatch;
import org.junit.Assert;
import org.junit.Test;
import util.Time;

public class LegacyWatchEquipmentCommandsTest {
    @Test
    public void dispatch_returnsFalse_forNonOwnedCommand() {
        TestFixture fixture = new TestFixture();
        try {
            boolean handled = fixture.handler.dispatch(
                fixture.computer,
                new ApplicationData("not-owned", null, 0, fixture.computer.ip),
                0
            );

            Assert.assertFalse(handled);
        } finally {
            fixture.close();
        }
    }

    @Test
    public void dispatch_fetchWatches_populatesPacketAssignment() {
        TestFixture fixture = new TestFixture();
        try {
            Watch watch = new Watch(fixture.computer);
            watch.setType(Watch.PETTY_CASH);
            watch.setOn(true);
            watch.setNote("alarm");
            watch.setCPUCost(2.5f);
            watch.setSearchFireWall(3);
            watch.setQuantity(125.0f);
            watch.setPort(9);
            fixture.computer.MyWatchHandler.addWatch(watch);
            fixture.computer.systemChange = false;

            boolean handled = fixture.handler.dispatch(
                fixture.computer,
                new ApplicationData("fetchwatches", null, 0, fixture.computer.ip),
                0
            );

            Assert.assertTrue(handled);
            Assert.assertTrue(fixture.computer.systemChange);
            PacketWatch[] packetWatches = fixture.computer.PA.getPacketWatches();
            Assert.assertEquals(1, packetWatches.length);
            Assert.assertEquals(Watch.PETTY_CASH, packetWatches[0].getType());
            Assert.assertEquals("alarm", packetWatches[0].getNote());
            Assert.assertTrue(packetWatches[0].getOn());
            Assert.assertEquals(9, packetWatches[0].getPort());
        } finally {
            fixture.close();
        }
    }

    @Test
    public void dispatch_changeWatchType_updatesWatchAndQueuesRefresh() {
        TestFixture fixture = new TestFixture();
        try {
            Watch watch = new Watch(fixture.computer);
            watch.setType(Watch.HEALTH);
            fixture.computer.MyWatchHandler.addWatch(watch);
            fixture.computer.Tasks.clear();

            boolean handled = fixture.handler.dispatch(
                fixture.computer,
                new ApplicationData("changewatchtype", new Integer[]{0, Watch.SCAN}, 0, fixture.computer.ip),
                0
            );

            Assert.assertTrue(handled);
            Assert.assertEquals(Watch.SCAN, watch.getType());
            Assert.assertEquals(1, fixture.computer.Tasks.size());
            ApplicationData queued = (ApplicationData) fixture.computer.Tasks.get(0);
            Assert.assertEquals("fetchwatches", queued.getFunction());
        } finally {
            fixture.close();
        }
    }

    private static final class TestFixture {
        private final Time time = new Time();
        private final Computer computer = new Computer("10.0.0.1", null, time, -1, null);
        private final LegacyWatchEquipmentCommands handler = new LegacyWatchEquipmentCommands();

        private TestFixture() {
            stopComputerThread(computer);
            computer.Tasks.clear();
        }

        private void close() {
            stopComputerThread(computer);
            time.clean();
        }
    }

    private static void stopComputerThread(Computer computer) {
        computer.run = false;
        Thread thread = computer.MyThread;
        if (thread != null) {
            thread.interrupt();
            try {
                thread.join(200);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        }
        computer.MyThread = null;
    }
}

File: src/GameServer/test/game/LegacyWatchEquipmentCommandsTest.kt
package game

import assignments.PacketWatch
import com.hackwars.rpc.ChangeWatchType
import com.hackwars.rpc.FetchPorts
import com.hackwars.rpc.FetchWatches
import com.hackwars.rpc.SetWatchOnOff
import com.hackwars.rpc.SetWatchObservedPorts
import game.payload.RequestEquipmentPayload
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertTrue
import org.junit.Test

class LegacyWatchEquipmentCommandsTest {
    @Test
    fun dispatch_returnsFalse_forNonOwnedCommand() {
        val fixture = TestFixture()
        try {
            val handled = fixture.handler.dispatch(
                fixture.computer,
                ApplicationData(FetchPorts(fixture.computer.ip), 0, fixture.computer.ip),
                0
            )

            assertFalse(handled)
        } finally {
            fixture.close()
        }
    }

    @Test
    fun dispatch_fetchWatches_populatesPacketAssignment() {
        val fixture = TestFixture()
        try {
            val watch = Watch(fixture.computer)
            watch.type = (Watch.PETTY_CASH)
            watch.on = (true)
            watch.note = ("alarm")
            watch.actualCpuCost = (2.5f)
            watch.searchFireWall = (3)
            watch.quantity = (125.0f)
            watch.port = (9)
            fixture.computer.MyWatchHandler.addWatch(watch)
            fixture.computer.systemChange = false

            val handled = fixture.handler.dispatch(
                fixture.computer,
                ApplicationData(FetchWatches(fixture.computer.ip), 0, fixture.computer.ip),
                0
            )

            assertTrue(handled)
            assertTrue(fixture.computer.systemChange)
            val packetWatches: Array<PacketWatch> = fixture.computer.PA.getPacketWatches()
            assertEquals(1, packetWatches.size)
            assertEquals(Watch.PETTY_CASH, packetWatches[0].getType())
            assertEquals("alarm", packetWatches[0].getNote())
            assertTrue(packetWatches[0].getOn())
            assertEquals(9, packetWatches[0].getPort())
        } finally {
            fixture.close()
        }
    }

    @Test
    fun dispatch_changeWatchType_updatesWatchAndQueuesRefresh() {
        val fixture = TestFixture()
        try {
            val watch = Watch(fixture.computer)
            watch.type = (Watch.HEALTH)
            fixture.computer.MyWatchHandler.addWatch(watch)
            fixture.computer.clearPendingTasks()

            val handled = fixture.handler.dispatch(
                fixture.computer,
                ApplicationData(ChangeWatchType(fixture.computer.ip, 0, Watch.SCAN), 0, fixture.computer.ip),
                0
            )

            assertTrue(handled)
            assertEquals(Watch.SCAN, watch.type)
            assertQueuedFetchWatches(fixture)
        } finally {
            fixture.close()
        }
    }

    @Test
    fun dispatch_setWatchOnOff_turnsWatchOnAndQueuesRefreshWhenWithinLimits() {
        val fixture = TestFixture()
        try {
            val watch = Watch(fixture.computer)
            watch.type = (Watch.PETTY_CASH)
            watch.actualCpuCost = (3.0f)
            watch.on = (false)
            fixture.computer.MyWatchHandler.addWatch(watch)
            fixture.computer.clearPendingTasks()

            val handled = fixture.handler.dispatch(
                fixture.computer,
                ApplicationData(SetWatchOnOff(fixture.computer.ip, 0, true), 0, fixture.computer.ip),
                0
            )

            assertTrue(handled)
            assertTrue(watch.on)
            assertQueuedFetchWatches(fixture)
        } finally {
            fixture.close()
        }
    }

    @Test
    fun dispatch_setWatchObservedPorts_updatesObservedPortsAndQueuesRefresh() {
        val fixture = TestFixture()
        try {
            val watch = Watch(fixture.computer)
            fixture.computer.MyWatchHandler.addWatch(watch)
            fixture.computer.clearPendingTasks()

            val handled = fixture.handler.dispatch(
                fixture.computer,
                ApplicationData(SetWatchObservedPorts(fixture.computer.ip, 0, arrayOf(3, 7, 9)), 0, fixture.computer.ip),
                0
            )

            assertTrue(handled)
            assertEquals(3, watch.observedPorts.size)
            assertEquals(3, watch.observedPorts[0])
            assertEquals(7, watch.observedPorts[1])
            assertEquals(9, watch.observedPorts[2])
            assertQueuedFetchWatches(fixture)
        } finally {
            fixture.close()
        }
    }

    @Test
    fun dispatch_requestEquipment_populatesPrimaryAndSecondaryDirectories() {
        val fixture = TestFixture()
        try {
            val agpCard = fixture.createEquipmentFile(HackerFile.AGP, "Alpha AGP")
            fixture.computer.MyFileSystem.addFile(agpCard, false)
            fixture.computer.systemChange = false

            val handled = fixture.handler.dispatch(
                fixture.computer,
                ApplicationData(RequestEquipmentPayload(13), 0, fixture.computer.ip),
                0
            )

            assertTrue(handled)
            assertTrue(fixture.computer.systemChange)
            assertNotNull(fixture.computer.PA.getDirectory())
            assertEquals(13, fixture.computer.PA.getDirectory()[0])
            assertNotNull(fixture.computer.PA.getSecondaryDirectory())
            assertEquals(13, fixture.computer.PA.getSecondaryDirectory()[0])
            val returned = fixture.computer.PA.getDirectory()[1] as HackerFile
            assertEquals("Alpha AGP", returned.getName())
            assertNotNull(returned.getContent()["bonusdata"])
        } finally {
            fixture.close()
        }
    }

    private fun assertQueuedFetchWatches(fixture: TestFixture) {
        val queuedTasks = fixture.computer.snapshotPendingTasks()
        assertEquals(1, queuedTasks.size)
        val queued = queuedTasks[0] as ApplicationData
        assertEquals("fetchwatches", queued.command.wireName())
    }

    private class TestFixture {
        val computer = Computer("10.0.0.1", null, -1, null)
        val handler = LegacyWatchEquipmentCommands()

        init {
            stopComputerRuntime(computer)
            computer.clearPendingTasks()
        }

        fun close() {
            stopComputerRuntime(computer)
        }

        fun createEquipmentFile(type: Int, name: String): HackerFile {
            val file = HackerFile(type)
            file.setLocation("")
            file.setName(name)
            val content = HashMap<Any?, Any?>()
            content["attribute0"] = "0"
            content["quality0"] = "0"
            content["attribute1"] = "1"
            content["quality1"] = "0"
            file.setContent(content)
            return file
        }
    }

    companion object {
        private fun stopComputerRuntime(computer: Computer) {
            computer.shutdown()
            computer.joinBlocking()
        }
    }
}
